### PR TITLE
Add s3 sts proxy support

### PR DESCRIPTION
**Pull request by phoebusm**
#### What does this implement or fix?
Add proxy support to AWS STS authentication

#### Any other comments?
`Aws::Auth::STSProfileCredentialsProvider` doesn't support proxy natively because proxy configuration hasn't been passed.
However, a manual sts client factory, which has the proxy config loaded, can be passed in the consturctor, to bring proxy support to the binaries.

End to end test will be added in a seperate PR.

Evidence:
```
(venv38) root@b3bc8b36aa29:/tmp/host# http_proxy="http://127.0.0.1:3128/" https_proxy="http://127.0.0.1:3128/" curl google
.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>
(venv38) root@b3bc8b36aa29:/tmp/host# curl google.com
curl: (7) Failed to connect to google.com port 80 after 25 ms: Connection refused
(venv38) root@b3bc8b36aa29:/tmp/host# http_proxy="http://127.0.0.1:3128/" https_proxy="http://127.0.0.1:3128/" ipython
Python 3.8.20 (default, Sep  7 2024, 18:35:08)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.3 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import arcticdb as adb

In [2]: ac = adb.Arctic('s3://s3.eu-north-1.amazonaws.com:BUCKET?aws_auth=sts&aws_profile=PROFILE&path_prefix=PREFIX')
20241216 16:14:08.540374 7083 I arcticdb | S3 proxy set from env var '127.0.0.1'
20241216 16:14:08.540681 7083 I arcticdb | S3 proxy set from env var '127.0.0.1'

In [3]: ac.list_libraries()
Out[3]: []
```
#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

**Review by poodlewars — COMMENTED** [review]
Is it possible to write a unit test so we check the right proxy config is getting passed down?
I assume that you've tested this manually?

File: cpp/arcticdb/storage/s3/s3_storage.cpp [hunk 1]
```diff
@@ -100,12 +101,20 @@ S3Storage::S3Storage(const LibraryPath &library_path, OpenMode mode, const S3Set
...
+            auto sts_config = get_proxy_config(conf.https() ? Aws::Http::Scheme::HTTPS : Aws::Http::Scheme::HTTP);
+            auto allocation_tag = Aws::STS::STSClient::GetAllocationTag();
+            sts_client_ = std::make_unique<Aws::STS::STSClient>(creds, Aws::MakeShared<Aws::STS::Endpoint::STSEndpointProvider>(allocation_tag), sts_config);
+            return sts_client_.get();
+        };
         auto cred_provider = Aws::MakeShared<Aws::Auth::STSProfileCredentialsProvider>(
                                 "DefaultAWSCredentialsProviderChain", 
                                 conf.aws_profile(), 
-                                std::chrono::minutes(static_cast<size_t>(ConfigsMap::instance()->get_int("S3Storage.STSTokenExpiryMin", 60)))
+                                std::chrono::minutes(static_cast<size_t>(ConfigsMap::instance()->get_int("S3Storage.STSTokenExpiryMin", 60))),
```
Comment (poodlewars) [inline cpp/arcticdb/storage/s3/s3_storage.cpp:114]: What does this do? Is it documented properly?

Comment (phoebusm) [reply cpp/arcticdb/storage/s3/s3_storage.cpp:114]: This is the token expiry setting. I have added this for the pending sts token expiry unit test
I think this shouldn't be useful to user so I specifically didn't put it to the manual

**Comment by phoebusm** [comment]
> Is it possible to write a unit test so we check the right proxy config is getting passed down? I assume that you've tested this manually?

The manual end to end test was done and I have pasted the evidence in the description. Adding the test to the CI will require some extra efforts.
I am not sure how to create the end to end test to check whether the right proxy is being passed down. For now, maybe a test to use sts with proxy but the proxy configuration is being set to a non-existing endpoint? This should show the proxy is being set

**Comment by phoebusm** [comment]
> > Is it possible to write a unit test so we check the right proxy config is getting passed down? I assume that you've tested this manually?
> 
> The manual end to end test was done and I have pasted the evidence in the description. Adding the test to the CI will require some extra efforts. I am not sure how to create the end to end test to check whether the right proxy is being passed down. For now, maybe a test to use sts with proxy but the proxy configuration is being set to a non-existing endpoint? This should show the proxy is being set

It's a dead end. The assertion check in the sdk makes `pytest.exception(...)` check impossible. 
The STS client passed is hidden under layers of classes and the config itself is private. It won't be reachable even with a dedicated C++ test, without some hacking

**Review by poodlewars — APPROVED** [review]

**Review by G-D-Petrov — APPROVED** [review]

**Pull request merged**